### ISSUE TITLE
Joint Consensus

### ISSUE THREAD

**Issue by Hoverbear** [tikv/tikv#3314]
# Design Prerequisites

## Demand and Design

Currently adding and removing nodes from a TiKV cluster is orchestrated by a PD cluster. We support adding as well as removing nodes, but we do not support a replace functionality (Eg cluster `ABC` becomes `BCD`.) Currently, this operation can be done unsafely. It is possible to do things such as add a node so that the majority of nodes are no longer up to date, causing progress to pause.

## Goal and Constraint

To be able to add, remove, and replace nodes fearlessly while knowing the cluster will transition safely.

For now, supporting `add_node`, `remove_node`, and maybe `replace_node` is enough. We do not need to support `add_nodes([1,2,3,4,5])` or other operations.

## Related Tests and Research

For a description of this problem and solution, please review **section 6** of https://ramcloud.stanford.edu/wiki/download/attachments/11370504/raft.pdf.

Someone wrote this analysis of the content: https://github.com/peterbourgon/raft/blob/master/JOINT-CONSENSUS.md

# Design Details

## Current Process Overview

The current process (for a user) is detailed here https://pingcap.com/docs/op-guide/horizontal-scale/#tikv. (This may be lacking some detail)

* `pd` receives a command to remove a node, or a TiKV node informs `pd` it needs to be added.
* `pd` then takes this policy and applies it to the cluster unilaterally.
* `pd` goes and instructs the peers to remove/add the other peer from their members.
* If the node is added, it begins as a learner.

## Joint Consensus Process Overview

* Leader of cluster receives a configuration change request.
* Enter transitional "Joint Consensus" state.
* Stores the `C(old, new)` as a log entry.
* Replicates `C(old, new)` entry.
* Once each peer adds the `C(old, new)` entry to the log:
    + It uses this `C(old,new)` for all future decisions.
    + In the case of a failure the leader may be elected using the configuration of `C(old)` or `C(old, new)` (depends on the new leader)
    + Peers in `C(new)` cannot make unilateral decisions at this point.
* Once `C(old, new)` is committed neither `C(old)` or `C(new)` can make decisions without the approval of the majority of the other group.
    + Only servers with `C(old, new)` can be elected as leader at this point due to the Leader Completeness property.
* The leader (which may have changed) inserts the `C(new)` log entry.
* Once each peer adds `C(new)` entry to the log:
    + Just as before, it immediately uses this configuration.
    + In case of failure the new leader will either be using `C(old, new)` or `C(new)`.
* Once the new `C(new)` has been committed the old configuration is no longer needed and the related servers can be shut down.

It is very important that at no point can `C(old)` or `C(new)` make unilateral decisions during the process.

* [ ] Test that `C(old)`/`C(new)` cannot make unilateral decisions when `C(new)`/`C(old)` is isolated and unable to make progress.

The Joint Consensus State:

* Log entries are still replicated to all servers.
* No leadership restrictions
* Agreement for both elections and entry commits requires **separate** majorities from both `C(new)` and `C(old)` clusters.
* Still services requests.
* Individual peers transition at different times.

## Already Present

The following things are already present:

* [x] **Raft:** `AppendEntries` RPCs can communicate configuration changes.
    * [x] Peers are serializable. (They are u64s)
    * [x] There is a `EntryType` value in `Entry` and a `EntryType::ConfChange` variant.
    * [x] There is a `ConfChangeType` enum with `AddNode`, `RemoveNode`, and `AddLearnerNode`.
    * [x] The `ConfChange` variant carries a `node_id` filed that can represent the node to be changed.
* [x] **Raft:** Can add and remove nodes.
    * [x] There is `add_node()`, `remove_node()`, and `add_learner()` functions. (though they are not using joint consensus)
    * [x] These methods are tested for functionality. (That they add/remove)
    * [x] There exists an `apply_conf_change` method.

Some of the features required for implementing joint consensus are already present in some form. The amount changes required are largely dependent on how we tackle the following two problems.

## Problem 1: User Facing Changes

There are a few options we have around how this change looks to the user. Namely how they add, remove, and replace nodes.

### Solution 1: Minimal Implementation

This solution attempts to change the user-facing system as little as possible. It would introduce the least changes.

* `pd-ctl` still manages nodes via `store remove $ID` or by receiving a new node request.
* `tikv-ctl` receives no changes.
* `pd` receives these commands and proxies them to the TiKV cluster, which handles them via joint consensus.
* `tikv-server`s take the command and relay it to the rafts.

### Solution 2: Migrate the Command

This solution changes the user-facing system so that `tikv-ctl` handles membership changes. `pd-ctl` no longer has the `store remove` command, and `tikv-server` nodes are added by the leader after being informed of their existence by `pd-server`.

* `pd-ctl` no longer has a `store remove $ID` command. `pd` still receives the message that a node is created, it forwards this to the TIKV leader.
* `tikv-ctl` now has a `store remove $ID` and possibly a `store replace` command.
* `pd` loses its responsibility of removing nodes (and possibly adding them), but it is still informed of these details.
* `tikv-server`s take the command and relay it to the rafts as needed.

### Solution 3: Have both

This solution combines the two by adopting the second solution but allowing `pd-ctl` to still add/remove nodes unilaterally.

There is a risk here of the users not understanding the difference between the commands issued to `pd-ctl` and `tikv-ctl`. For example they may issue `remove`  I feel like this solution is dangerous for that reason.

## Problem 2: Features

One of the key things we'd like to do with this feature is to be able to *replace* nodes safely. This can be done via `remove_node` and then `add_node`, however there is likely some optimization around this to avoid having to go through the joint consensus twice (once for removal, once for addition).

Additionally, Joint Consensus describes the process of adding and removing arbitrary number of nodes. (A replace is an add and a remove at the same time.) Currently we only support adding and removing one node at at time.

### Solution 1: One at a Time, Without Replace

This solution is the least optimal, but it is also the simplest. The amount of nodes to change at a time remains one, and there is no replace. All this solution would do is to push the existing membership changes into TiKV itself.

This solution would likely not cause compatibility issues as it does not require any changes to protobufs. Testing around removing outdated nodes would still need to be done though.

### Solution 2: One at a Time, With Replace

This solution supports an explicit "replace node x with node y". This would do the following:

* Create a conf change which adds the node
* Once this conf change (using the joint consensus algorithm described) is applied create a conf change that removes the other node.
* Once applied the process is done.

This solution will likely not be backwards compatible, protobufs would change and a new command (replace) would not be actionable by outdated nodes.

### Solution 3: Many-at-a-time, No explicit Replace

This solution uses a new membership change call which supports injecting a new node list distributed through the leader. Sending a node list of `[u64]` in this new call allows us to calculate:

* `sent_nodes = C(old)`
* `sent_nodes = C(new)`
* `union(current_nodes, sent_nodes) = C(old, new)`
* `current_nodes - sent_nodes` the nodes to be removed.
* `sent_nodes - current_nodes` the nodes to be added.

This does not explicitly support replace. This solution supports using add and remove in the same change, so a transition from `ABC` to `BCD` is seamless, which allows for replace functionality.

Since this solution introduces a new call that operates without `pd` the existing code can either be left in (as a force mechanism), removed, or altered to call the new "set nodes" operation on the leader from `pd`.

This solution would introduce or change the most code and require the most testing.

It’s currently unclear how we would enqueue the `C(new)` configuration change. The mechanism described in the paper would only append the `C(new)` **after** the `C(old, new)` is committed.

This is important, if we added `C(new)` right away after the `C(old, new)` entry we could reduce cluster stability. So instead the log entry for `C(old, new)` should be able to emit a `C(new)` which can be appended when `C(old, new)` is committed.

In order to accomplish this (creating a new append when committing an entry) I think we can use `check_message_with_progress` which is already handling `MessageType::MsgTransferLeader`.

The functionality and flow to reach the `apply_conf_change` function in the raft library currently is not the same as what we need. We need to use `C(old,new)` after receiving the message (in `step`).

## Potential Issues
These are possible issues which joint consensus is careful to cover. Using these as test cases can help is verify correctness.

### Issue: Catching up new nodes

If a node is added without any log entries it may take quite some time before it catches up, which could potentially render the cluster unable to make progress.

This is resolved via the learner state, when nodes in this state are not considered for majorities.

* [ ] Test that ensures this does not cause disruption. It should:
    + Create a cluster of one, let it elect itself leader.
    + Add two nodes that have no log, both should enter as learners.
    + Ensure the new nodes are learners.
    + Ensure the cluster remains operational until the nodes are no longer learners.

### Issue: Leader not in `C(new)`

It is possible that the peer which is currently leader in `C(old)` will not be part of `C(new)`. This means at some point during `C(old, new)` the leader will be managing a cluster it is not a part of until the `C(new)` entry is committed.

During this time it replicates log entries, but does not count itself in majorities. When `C(new)` is committed a leadership transfer can occur to a new node in `C(new)`, otherwise only nodes from `C(old)` may be chosen.

* [ ] Test a normal transition with a leader in both old and new cluster.
    * Create a cluster of (three/five) let it elect `1` as leader.
    * Add a node (`4`/`6`) and remove one (`1`).
    * Ensure that during the joint the leader remains `1`.
    * Ensure that after the joint `1` steps down.
    * Ensure that `1` is removed from the cluster.
* [ ] Test a transition that includes a leader failure when it is to be removed.
    * Add a node (`4`/`6`) and remove one (`1`).
    * Ensure that during the joint the leader remains `1`.
    * Cause a failure on `1` so it cannot progress.
    * Ensure that during the joint a new node from the original cluster is elected. (NOT the new one)
    * Ensure that `1` is removed from the cluster. (Since there is still a majority this should be possible)

### Issue: Removed servers can disrupt the cluster

When a server is removed from the cluster it will no longer receive heartbeats or other messages, and will eventually timeout. Once that happens they will try to start a new election with a new term. This will cause the leader of the `C(new)` to step down. Though this will cause a new election, it will result in decreased cluster availability.

In order to prevent this, peers will (TODO: no longer service, or reject?) `RequestVote` RPC they receive within the minimum election timeout of hearing from their current leader. They will not update their term or grant a vote.

This way, so long as the leader is able to get heartbeats to nodes there will be no disruption.

* [ ] Test that if a peer receives a `RequestVote` RPC immediately after hearing from their current leader that they will not update their term or grant a vote.

## Open Questions

* What will change about the membership change process in relation to `pd` and `tikv`?
    * Currently `pd` manages configuration changes.
        * When a node is added `pd` is informed and it issues the configuration change.
        * To remove a node `pd-ctl` is used with `store delete $ID`.
    * After this it will be possible to move this functionality into `tikv-ctl` and have the `pd-ctl` command just send the command to the leader (for backwards compatibility).
* Is it possible to create a situation where the cluster is in a state where it does not have quorum during a combined add/remove action?
* How to write `C(old,new)` to the log is obvious, when the action is started. I still need to determine the correct mechanism to set up the appending of `C(new)` once `C(old,new)` is replicated.

**Comment by Hoverbear** [comment]
I currently prefer solution 2 to problem 1, and solution 3 to problem 2.

**Comment by BusyJay** [comment]
Now we are trying port Etcd's implement. Close it now, thanks!
